WalletWasabi/WabiSabi/Crypto/CredentialIssuer.cs

```diff
@@ -193,34 +214,67 @@ public ICommitableCredentialsResponse PrepareResponse(CredentialsRequest registr
 			var transcript = BuildTransnscript(registrationRequest.IsNullRequest);
 
 			// Verify all statements.
-			var areProofsValid = ProofSystem.Verify(transcript, statements, registrationRequest.Proofs);
-			if (!areProofsValid)
+			try
 			{
-				throw new WabiSabiCryptoException(WabiSabiCryptoErrorCode.CoordinatorReceivedInvalidProofs);
+				var areProofsValid = ProofSystem.Verify(transcript, statements, registrationRequest.Proofs);
+				if (!areProofsValid)
+				{
+					throw new WabiSabiCryptoException(WabiSabiCryptoErrorCode.CoordinatorReceivedInvalidProofs);
+				}
 			}
+			catch
```

Comment:
Don't use `try-catch` for control flow. You can just move all what is in the `catch` block into the `if (!areProofsValid)` before throwing the exception.

Reply:
this is in case the `Guard` triggers, or `ProofSystem` throws some other kind of exception

Reply:
No method can work with invalid inputs and that's why we have to validate what we receive from the users  _before_ doing anything else. We already discussed this topic and agreed to go back to it later when I proposed this solution: https://github.com/lontivero/Wasabito/commit/03df012c578db4901a1e36c44a7365539f21a0d1

Perhaps the time is now. We can do with this solution which is one of the standard ways to implement model validations in asp.net or any other way but the idea is always the same: to validate what we receive from the outside.

Reply:
I don't see what's wrong with cleaning up in a catch block, independently of such validation, which should be done in a separate PR?

The point is that if we miss anything, an attacker can't just fill up server memory with invalid serial numbers. Specifically here this should be `CredentialIssuer`'s responsibility to make sure the number of statements (which it calculates independently) is equal to the number of expected proofs, so we can't validate this input before the credential issuer constructs the statement, but even so writing it in a robust way seems more correct to me, it's not like we have access to formal verification to allow us to ignore this kind of edge case.

Reply:
put another way, can we statically guarantee that `ProofSystem` will never throw any exception? if not then it's not OK for `CredentialIssuer` to rely on that assumption preventing the nullifier set from clogging

Reply:
Another way to write Yuval's code which I find more idiomatic and common in C# is:

```csharp
bool areProofsValid = false;

try
{
  // Verify all statements.
  areProofsValid = ProofSystem.Verify(transcript, statements, registrationRequest.Proofs);
  if (!areProofsValid)
  {
    throw new WabiSabiCryptoException(WabiSabiCryptoErrorCode.CoordinatorReceivedInvalidProofs);
  }
}
finally
{
  if (!areProofsValid) 
  {
    // Request was invalid, but all serial numbers were unused.
    // Ensure nullifier set can't be clogged with invalid serial
    // numbers. Valid serial numbers are only issued in response to
    // valid requests, which in turn depend on valid ownership
    // proofs and the banning mechanism, and they actually matter
    // for double spending, so only they need to be stored in the
    // nullifier set.
    lock (SerialNumbersLock)
    {
      foreach (var serialNumber in presentedSerialNumbers)
      {
        SerialNumbers.Remove(serialNumber);
      }
    }  
  }
}
```

Reply:
I suggest deciding on the validation in the next PR and finish with this.

Reply:
@lontivero are you happier with Kiminuo's suggestion?

Reply:
I think it is better, yes.